orttraining/orttraining/training_ops/cuda/communication/recv.cc

```diff
@@ -4,74 +4,112 @@
-      kCudaExecutionProvider,                                            \
-      KernelDefBuilder()                                                 \
-          .InputMemoryType<OrtMemTypeCPUInput>(0)  /* CPU variable */    \
-          .OutputMemoryType<OrtMemTypeCPUOutput>(0)  /* CPU variable */  \
-          .TypeConstraint("T", DataTypeImpl::GetTensorType<T>())         \
-          .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>()), \
-      Recv<T>);
-
-REGISTER_RECV_KERNEL_TYPED(MLFloat16)
-REGISTER_RECV_KERNEL_TYPED(float)
-REGISTER_RECV_KERNEL_TYPED(double)
+ONNX_OPERATOR_KERNEL_EX(
+    Recv,
+    kMSDomain,
+    1,
+    kCudaExecutionProvider,
+    KernelDefBuilder()
+        .InputMemoryType<OrtMemTypeCPUInput>(0)    /* CPU variable */
+        .InputMemoryType<OrtMemTypeCPUInput>(1)    /* CPU variable */
+        .OutputMemoryType<OrtMemTypeCPUOutput>(0)  /* CPU variable */
+        .TypeConstraint("TInt64", DataTypeImpl::GetTensorType<int64_t>())
+        .TypeConstraint("TBool", DataTypeImpl::GetTensorType<bool>()), 
+    Recv);
+
 
 void CUDART_CB HostRecv(void* args) {
   CommInfo_t* info = reinterpret_cast<CommInfo_t*>(args);
   int mpi_code  = MPI_Recv(info->buffer, info->size, MPI_CHAR, info->rank, info->tag, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
   ORT_ENFORCE(mpi_code == MPI_SUCCESS, "MPI Recv fails.");
 }
 
-template <typename T>
-Status Recv<T>::ComputeInternal(OpKernelContext* ctx) const {
+Status Recv::ComputeInternal(OpKernelContext* ctx) const {
+  // Extract Remote rank
+  const Tensor* remote_rank_tensor = ctx->Input<Tensor>(0);
+  const int64_t* remote_rank = remote_rank_tensor->template Data<int64_t>();
+  int src = static_cast<int>(*remote_rank);
+
   // Check if control signal is true.
-  const Tensor* input_signal_tensor = ctx->Input<Tensor>(0);
+  const Tensor* input_signal_tensor = ctx->Input<Tensor>(1);
   const bool* input_signal = input_signal_tensor->template Data<bool>();
   ORT_ENFORCE(*input_signal, "Input control signal of Recv must be true before executing the node.");
 
-  // Start the communication.
+  // Create buffers
+  int tensor_num = static_cast<int> (element_types_.size());
+  // TODO move the following variables to member variables for extending life-time
+  // if we want to make the entire call async
+  std::vector<size_t> prefix_tensor_shape_sizes(tensor_num);
+  std::vector<int64_t> aggregated_tensor_shapes;
+  size_t aggregated_aligned_tensor_bytes = 0;
+  size_t alignment = 256;
+
+  // Start communication
   int world_rank;
   MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
+  ORT_ENFORCE(world_rank != src, "Receive data from rank ", src, " on the rank ", world_rank, ".");
 
-  ORT_ENFORCE(world_rank == dst_, "Rank ", world_rank, " is reciving data but the expected reciving rank is ", dst_, ".");
 
-  cudaStream_t commStream;
+  // Enqueue communication functions to a GPU stream.
+  // Keep Wei-Sheng's local stream
```

Comment:
Nit: since orttraing is now public, it's better leave out dev's name in comment.

Reply:
Change to todo + id, like other repo doing?